src/Streamly/Internal/Data/Stream/StreamK.hs

```diff
@@ -638,6 +639,19 @@ takeWhile p = go
                        | otherwise = stp
          in foldStream st yieldk single stp m1
 
+{-# INLINE takeEndBy #-}
+takeEndBy :: (a -> Bool) -> Stream m a -> Stream m a
+takeEndBy p = go Nothing
+    where
+    go Nothing m1 = mkStream $ \st yld sng stp ->
+        let single a   | not $ p a = sng a
+                       | otherwise = stp
+            yieldk a r | not $ p a = yld a (go Nothing r)
+                       | otherwise = yld a (go (Just a) r)
```

Comment:
This code is incorrect. 

There is no need to have the `Maybe` argument in `go`.

The code should be:

```
            yieldk a r | not $ p a = yld a (go r)
                       | otherwise = sng a
```